Change interfaces & unions to require enums

phated opened the pull request:
@sunli829 this is my first time working on a macro, so it might be bad.

I have found it weird that the `#[async_graphql::Interface]` macro replaces a `struct` definition with a `enum` definition. You can see me trying to extend an interface with diesel traits in https://github.com/phated/twentyfive-stars/blob/master/server/src/data/character_mode.rs - and you can see the "weird" behavior at https://github.com/phated/twentyfive-stars/blob/master/server/src/data/character_mode.rs#L116

By making this macro only usable on enums, we have a more natural usage pattern when extending it.

sunli829 commented:
Why not call `into` directly?

```rust
UpgradeMode {
          id,
          title,
          stars,
          type_,
          faction,
          traits,
          attack_modifier,
          defense_modifier,
        }.into()
```

phated commented:
For the creation, `.into()` would work, but I am working on a new feature where I need to pattern match on the enum variants.

If you agree to this implementation, I will update the documentation and also make changes to Union. Please let me know

phated commented:
This adds some extra verbosity in the definition, but I believe it makes the code easier to understand when you need to use the data types in other parts of the server.

sunli829 commented:
Someone really asked me why `Interface` would be a struct, and I told him it was actually an enumeration. In fact, when I first implemented the `Interface`, it was an enumeration, but I was lazy. I did n’t like to write any extra code. The enumeration forced me to make a repeated name, like this `enum MyInterface {Foo (Foo)}` , so changed to struct. But it does make the code difficult to understand. Thanks for your PR, I will merge later, because I need to change the `Union` to use enumeration.

phated commented:
Great! I will work on `Union` now!

sunli829 commented:
Although you have a lot of open source responsibilities, I still want to add you to the collaborators, because your suggestions are very important to this project, are you willing? 😁

phated commented:
> are you willing?

Yeah, I am happy to help where I can. Thank you! 🙇 I "watch" this project already, so I see all of the activity 😄

sunli829 commented:
Great, let's do the best GraphQL library of rust together. 😄

phated commented:
@sunli829 the Union is implemented. I also added some tests and documentation for it. In addition, I updated the book examples.

Do you need me to update the examples or anything else?

sunli829 commented:
Too fast, the code generation part I wrote is not very good, a little messy.😂
I will translate the book to Chinese, if you want, you can help update the Examples repo.

phated commented:
> Too fast, the code generation part I wrote is not very good, a little messy.😂

@sunli829 do you need me to change something? Or are you going to improve on this?

> I will translate the book to Chinese, if you want, you can help update the Examples repo.

The book doesn't mention `struct`, so I changed the example. I don't think anything else needs to be done. 

I will work on the examples repo when we finish this. Probably tomorrow because it is late for me.

Thank you!

sunli829 commented:
> do you need me to change something? Or are you going to improve on this?
Now it seems unnecessary, after all, the amount of code is not too large. 😁

> I will work on the examples repo when we finish this. Probably tomorrow because it is late for me.
Thanks, take a break early, good night. 😁

sunli829 commented:
I will update the examples repo after the new version is released. This version changes a lot. 😁
